### PR TITLE
chromium: added patch to support master_preferences

### PR DESCRIPTION
**Pull request by luis-caldas**
###### Description of changes

Added a patch to make chromium look for `master_preferences` in `/etc/chromium` as other distros do.

Added also module options for both the new `master_preferences` and recommended policies.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by RaitoBezarius** [comment]
Hm, could we use XDG standards for the lookup paths? Hardcoding `/etc/chromium` feels bad to me.

**Comment by luis-caldas** [comment]
Thats how other distros have been handling the issue: [Debian](https://sources.debian.org/patches/chromium/104.0.5112.101-1/debianization/master-preferences.patch/), [Fedora](https://src.fedoraproject.org/rpms/chromium/blob/rawhide/f/chromium-68.0.3440.106-master-prefs-path.patch).
We have been hardcoding these paths also on the module options.

File: nixos/modules/programs/chromium.nix [hunk 1]
```diff
@@ -74,6 +74,25 @@ in
           "https://encrypted.google.com/complete/search?output=chrome&q={searchTerms}";
       };
 
+      initialOpts = mkOption {
+        type = types.attrs;
+        description = ''
+          Default preferences to be set at first startup.
+          <link xlink:href="https://support.google.com/chrome/a/answer/187948?hl=en">https://support.google.com/chrome/a/answer/187948?hl=en</link>
+        '';
```
Comment (SuperSandro2000) [inline nixos/modules/programs/chromium.nix:82]:
```suggestion
        description = lib.mdDoc ''
          Default preferences to be set at first startup.
          <https://support.google.com/chrome/a/answer/187948>
        '';
```

File: nixos/modules/programs/chromium.nix [hunk 2]
```diff
@@ -96,20 +115,61 @@ in
           }
         '';
       };
+
+      recommendedOpts = mkOption {
+        type = types.attrs;
+        description = ''
+          Recommended chromium policy options.
+          Recommended policies can be changed by the user.
```
Comment (SuperSandro2000) [inline nixos/modules/programs/chromium.nix:123]:
```suggestion
          Recommended chromium policy options which can be changed by the user.
```

File: nixos/modules/programs/chromium.nix [hunk 3]
```diff
@@ -96,20 +115,61 @@ in
...
+          {
+            "BrowserSignin" = 0;
+            "SyncDisabled" = true;
+            "PasswordManagerEnabled" = false;
+            "SpellcheckEnabled" = true;
+          }
+        '';
+      };
+
     };
```
Comment (SuperSandro2000) [inline nixos/modules/programs/chromium.nix:136]:
```suggestion
      };
    };
```

File: nixos/modules/programs/chromium.nix [hunk 4]
```diff
@@ -96,20 +115,61 @@ in
...
+        fullListFiles = map (
+          folder:
+            lib.mapAttrsToList (name: value: { name = "${folder}/${name}"; inherit value; }) commonStructure
+        ) baseFolders;
+      in builtins.listToAttrs (lib.flatten fullListFiles);
+
+    in allFiles;
+
   };
+
```
Comment (SuperSandro2000) [inline nixos/modules/programs/chromium.nix:174]:
```suggestion
  };
```

File: nixos/modules/programs/chromium.nix [hunk 5]
```diff
@@ -96,20 +115,61 @@ in
...
+        "policies/recommended/default.json" = addText (builtins.toJSON cfg.recommendedOpts);
+      };
+
+      # Build full file list
+      allFiles = let
+        fullListFiles = map (
+          folder:
+            lib.mapAttrsToList (name: value: { name = "${folder}/${name}"; inherit value; }) commonStructure
+        ) baseFolders;
+      in builtins.listToAttrs (lib.flatten fullListFiles);
```
Comment (SuperSandro2000) [inline nixos/modules/programs/chromium.nix:169]:
```suggestion
      in listToAttrs (lib.flatten fullListFiles);
```

**Review by SuperSandro2000 — COMMENTED** [review]
Please follow the contributing guide when naming your commits. Otherwise LGTM.

**Review by primeos — CHANGES_REQUESTED** [review]
An important question from me as the current Chromium maintainer:
Are you willing to maintain that additional patch? I.e., I'll contact you if the patch breaks and you'll rebase it. I'm asking because I don't have the time and motivation for that and ~~would~~ will remove it before it delays a stable release.

In general I'm not keen on any additional Chromium patches but I'm fine with it as long as doesn't make maintaining Chromium even more time consuming.

Also: Please write proper commit messages (https://cbea.ms/git-commit/). Thanks :) The current ones don't follow our Nixpkgs conventions and you should also add a brief description/body (the content of the PR description should be acceptable but some more details/information would make it even better). And in common.nix I'd also mention why we need this patch (for the new NixOS module functionality and I'd also add a link to the GitHub issue there or in the commit message).

**Pull request closed without merging**